{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta(fastq):\n",
    "    SeqIO.convert(fastq, \"fastq\", fastq[:-6]+'.fasta', \"fasta\")\n",
    "    return fastq[:-6]+'.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_fasta('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bedoutput_ITS(reference_fasta, primer_fasta):\n",
    "    \"\"\"The purpose of this function is to generate a fasta file of the reference genomeregions between primers, \n",
    "    which should correspond with the region sequenced using rhe MinION. This requires first the generation of a \n",
    "    database using the nucleotide makeblastdb command, followed by running a blast search across the database \n",
    "    using the primer sequence. These command line commands are processed through subprocess. Subsequently, the \n",
    "    resulting outfile from the blast is read in as a dataframe and this dataframe is manipulated in order to \n",
    "    utilise only the minimum evalues. Each forward sequence is then compared to each reverse sequence, given they\n",
    "    are on the same contig and the resultant read would be within 2000-4000 bp long. These results are saved as \n",
    "    a bedfile for use in the get_bedfasta function\"\"\"\n",
    "    cmd = 'makeblastdb -in %s -dbtype nucl -out %s' % (reference_fasta, reference_fasta[:-6]+'_db')\n",
    "    subprocess.getoutput(cmd)\n",
    "    cmd2 = 'blastn -query %s -db %s -evalue=100000 -task \"blastn-short\" -outfmt 6 > %s' % (primer_fasta, reference_fasta[:-6]+'_db', reference_fasta[:-6]+'.outfmt6')\n",
    "    subprocess.getoutput(cmd2)\n",
    "    df = pd.read_csv(reference_fasta[:-6]+'.outfmt6', sep=\"\\t\", header=None, names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"])\n",
    "    forward_df = df.loc[(df['qseqid']=='forward')]\n",
    "    forward_df = forward_df.loc[df['evalue']==min(forward_df['evalue'])]\n",
    "    reverse_df = df.loc[(df['qseqid']=='reverse')]\n",
    "    reverse_df = reverse_df.loc[df['evalue']==min(reverse_df['evalue'])]\n",
    "    forward_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "    for i in range(0, len(forward_df)):\n",
    "        if forward_df['sstart'][i] <= forward_df['send'][i]:\n",
    "            forward_bed = forward_bed.append({'chrom': forward_df['sseqid'][i],'chromStart': (forward_df['sstart'][i])-1, 'chromEnd': forward_df['send'][i]}, ignore_index=True)\n",
    "        else:\n",
    "            forward_bed = forward_bed.append({'chrom': forward_df['sseqid'][i],'chromStart': (forward_df['send'][i])-1, 'chromEnd': forward_df['sstart'][i]}, ignore_index=True)\n",
    "    forward_bed.sort_values(['chromStart'])\n",
    "    reverse_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "    for i in reverse_df.index:\n",
    "        if reverse_df['sstart'][i] < reverse_df['send'][i]:\n",
    "            reverse_bed = reverse_bed.append({'chrom': reverse_df['sseqid'][i],'chromStart': (reverse_df['sstart'][i])-1, 'chromEnd': reverse_df['send'][i]}, ignore_index=True)\n",
    "        else:\n",
    "            reverse_bed = reverse_bed.append({'chrom': reverse_df['sseqid'][i],'chromStart': (reverse_df['send'][i])-1, 'chromEnd': reverse_df['sstart'][i]}, ignore_index=True)\n",
    "    reverse_bed.sort_values(['chromStart'])\n",
    "    intervals_list = []\n",
    "    for row in forward_bed.itertuples(index=True, name='Pandas'):\n",
    "        counter = 0\n",
    "        for rows in reverse_bed.itertuples(index=True, name='Pandas'):\n",
    "            # Here lies the difference between the ITS and TEF versions of this function - for ITS the read length range is 2000-4000\n",
    "            if getattr(row, 'chrom') == getattr(rows, 'chrom') and 2000 < np.absolute(getattr(row, 'chromEnd') - getattr(rows, 'chromStart')) < 4000:\n",
    "                counter += 1\n",
    "                intervals_list.append([forward_bed['chrom'][getattr(row, 'Index')], forward_bed['chromEnd'][getattr(row, 'Index')], reverse_bed['chromStart'][getattr(rows, 'Index')]])\n",
    "        if counter > 1:\n",
    "            print(\"WARNING\")\n",
    "    intervals_frame = pd.DataFrame(data=intervals_list, columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "    print(intervals_frame)\n",
    "    intervals_frame.to_csv(reference_fasta[:-6]+'.ITS.bedfile', sep='\\t', header=False, index=False)\n",
    "#     lines = []\n",
    "#     for i in range(0, len(intervals_frame)):\n",
    "#         lines.append(['%i\\t%i\\t%i' % (intervals_frame['chrom'][i], intervals_frame['chromStart'][i], intervals_frame['chromEnd'][i])]) \n",
    "#     with open(reference_fasta[:-6]+'.ITS.bedfile', 'w') as f:\n",
    "#         for i in range(0, len(lines)):\n",
    "#             f.writelines(lines[i])\n",
    "#             f.writelines('\\n')\n",
    "    return reference_fasta[:-6]+'.ITS.bedfile'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_bedoutput_ITS('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.reference.fasta', '/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/ITS_primers.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bedfasta(bedfile, reference_fasta):\n",
    "    cmd2 = 'bedtools getfasta -fo %s -fi %s -bed %s' % (reference_fasta[:-6]+'.ITS.bedoutput.fasta', reference_fasta, bedfile)\n",
    "    subprocess.getoutput(cmd2)\n",
    "    return reference_fasta[:-6]+'.ITS.bedoutput.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_bedfasta('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.reference.ITS.bedfile', '/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.reference.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def length_filter(fasta, minimum, maximum):\n",
    "    fasta_dict = SeqIO.to_dict(SeqIO.parse(fasta, \"fasta\"))\n",
    "    short_dict = fasta_dict.copy()\n",
    "    for key in fasta_dict:\n",
    "        if len(fasta_dict[key].seq) not in range(minimum, maximum + 1):\n",
    "            del short_dict[key]\n",
    "    SeqIO.write(short_dict.values(), fasta[:-6]+'.filtered_base_%s-%s.fasta' % (minimum, maximum), \"fasta\")\n",
    "    return fasta[:-6]+'.filtered_base_%s-%s.fasta' % (minimum, maximum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# length_filter('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.fasta', 2000, 4000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pcr_seqs(filtered_fasta, oligos):\n",
    "    tmp = SeqIO.to_dict(SeqIO.parse(filtered_fasta, \"fasta\"))\n",
    "    new_name = filtered_fasta+'.mothur5.%s.fasta' % (oligos[-10:-7])\n",
    "    SeqIO.write(tmp.values(), new_name, \"fasta\")\n",
    "    m = Mothur()\n",
    "    m.pcr.seqs(fasta=new_name, oligos=oligos, pdiffs=5, rdiffs=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pcr_seqs('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.filtered_base_2000-4000.fasta', '/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/ITS.oligos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pairwise_aln(reference_bed_output, mothur_query_fasta):\n",
    "    gen_dict = SeqIO.to_dict(SeqIO.parse(reference_bed_output, \"fasta\"))\n",
    "    print(gen_dict)\n",
    "    gen_dict_seqs = []\n",
    "    for key in gen_dict:\n",
    "        gen_dict_seqs.append(gen_dict[key].seq)\n",
    "    print(gen_dict_seqs)\n",
    "    q_dict = SeqIO.to_dict(SeqIO.parse(mothur_query_fasta, \"fasta\"))\n",
    "    error5w = {}\n",
    "    error5w_key_entries = ['keys', 'alnslengths', 'aap', 'acp', 'agp', 'atp', 'agapp', 'cap', 'ccp', 'cgp', 'ctp', 'cgapp', 'gap', 'gcp', 'ggp', 'gtp', 'ggapp', 'tap', 'tcp', 'tgp', 'ttp', 'tgapp', 'gapap', 'gapcp', 'gapgp', 'gaptp', 'expeca', 'expecc', 'expecg', 'expect']\n",
    "    for item in error5w_key_entries:\n",
    "        error5w[item] = []\n",
    "    for key in q_dict:\n",
    "        error5w['keys'].append(key)\n",
    "        alignments = pairwise2.align.globalms(gen_dict_seqs[0], q_dict[key].seq, 1, -1, -1, 0)\n",
    "        total = 0\n",
    "        aa = 0\n",
    "        ac = 0\n",
    "        ag = 0\n",
    "        at = 0\n",
    "        agap = 0\n",
    "        ca =0\n",
    "        cc=0\n",
    "        cg=0\n",
    "        ct=0\n",
    "        cgap=0\n",
    "        ga=0\n",
    "        gc=0\n",
    "        gg=0\n",
    "        gt=0\n",
    "        ggap=0\n",
    "        ta=0\n",
    "        tc=0\n",
    "        tg=0\n",
    "        tt=0\n",
    "        tgap=0\n",
    "        gapa=0\n",
    "        gapc=0\n",
    "        gapg=0\n",
    "        gapt=0\n",
    "        tmpa=0\n",
    "        tmpc=0\n",
    "        tmpg=0\n",
    "        tmpt=0\n",
    "        comparison = alignments[0][0:2]\n",
    "        for i in range(0, len(comparison[0])):\n",
    "            if comparison[0][i] == \"A\":\n",
    "                tmpa += 1\n",
    "                if comparison[1][i] == \"C\":\n",
    "                    ac+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    ag+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    at+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"A\":\n",
    "                    aa+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    agap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"C\":\n",
    "                tmpc+=1\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    ca+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    cg+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    ct+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    cc+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    cgap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"G\":\n",
    "                tmpg+=1\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    ga+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    gc+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    gt+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    gg+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    ggap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"T\":\n",
    "                tmpt+=1\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    ta+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    tc+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    tg+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    tt+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    tgap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"-\":\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    gapa+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    gapc+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    gapg+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    gapt+=1\n",
    "                    total += 1\n",
    "        error5w['acp'].append(100*ac/total)\n",
    "        error5w['agp'].append(100*ag/total)\n",
    "        error5w['atp'].append(100*at/total)\n",
    "        error5w['agapp'].append(100*agap/total)\n",
    "        error5w['aap'].append(100*aa/total)\n",
    "        error5w['cap'].append(100*ca/total)\n",
    "        error5w['cgp'].append(100*cg/total)\n",
    "        error5w['ctp'].append(100*ct/total)\n",
    "        error5w['cgapp'].append(100*cgap/total)\n",
    "        error5w['ccp'].append(100*cc/total)\n",
    "        error5w['gap'].append(100*ga/total)\n",
    "        error5w['gcp'].append(100*gc/total)\n",
    "        error5w['gtp'].append(100*gt/total)\n",
    "        error5w['ggapp'].append(100*ggap/total)\n",
    "        error5w['ggp'].append(100*gg/total)\n",
    "        error5w['tap'].append(100*ta/total)\n",
    "        error5w['tcp'].append(100*tc/total)\n",
    "        error5w['tgp'].append(100*tg/total)\n",
    "        error5w['tgapp'].append(100*tgap/total)\n",
    "        error5w['ttp'].append(100*tt/total)\n",
    "        error5w['gapap'].append(100*gapa/total)\n",
    "        error5w['gapcp'].append(100*gapc/total)\n",
    "        error5w['gapgp'].append(100*gapg/total)\n",
    "        error5w['gaptp'].append(100*gapt/total)\n",
    "        error5w['alnslengths'].append(total)\n",
    "        error5w['expeca'].append(100*tmpa/total)\n",
    "        error5w['expecc'].append(100*tmpc/total)\n",
    "        error5w['expecg'].append(100*tmpg/total)\n",
    "        error5w['expect'].append(100*tmpt/total)\n",
    "    error5w_frame = pd.DataFrame.from_dict(error5w, orient='index', columns=error5w['keys'])\n",
    "    error5w_frame = error5w_frame.transpose()\n",
    "    error5w_frame.to_csv(mothur_query_fasta[:-5]+reference_bed_output[-19:-15]+'dataframe.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pairwise_aln('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.reference.ITS.bedoutput.fasta', '/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.filtered_base_2000-4000.fasta.mothur5.ITS.pcr.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_stats_ind(filename):\n",
    "    \n",
    "    frame = pd.DataFrame.from_csv(filename)\n",
    "    result5w = {}\n",
    "    result5w_key_entries = ['astats', 'cstats', 'gstats', 'tstats', 'gapstats', 'expec', 'ostats']\n",
    "    for item in result5w_key_entries:\n",
    "        result5w[item] = []\n",
    "    result5w['astats'] = [frame['aap'], frame['acp'], frame['agp'], frame['atp'], frame['agapp']]\n",
    "    result5w['cstats'] = [frame['cap'], frame['ccp'], frame['cgp'], frame['ctp'], frame['cgapp']]\n",
    "    result5w['gstats'] = [frame['gap'], frame['gcp'], frame['ggp'], frame['gtp'], frame['ggapp']]\n",
    "    result5w['tstats'] = [frame['tap'], frame['tcp'], frame['tgp'], frame['ttp'], frame['tgapp']]\n",
    "    result5w['gapstats'] = [frame['gapap'], frame['gapcp'], frame['gapgp'], frame['gaptp']]\n",
    "    \n",
    "    fg, ax = plt.subplots(5, 5, figsize=(20,20))\n",
    "    fg.suptitle(\"Base Percentages\", fontsize=28, y=1.35)\n",
    "    fg.subplots_adjust(top=1.3)\n",
    "    titles = ['A-A', 'A-C', 'A-G', 'A-T', 'A-GAP', 'C-A', 'C-C', 'C-G', 'C-T', 'C-GAP', 'G-A', 'G-C', 'G-G', 'G-T', 'G-GAP', 'T-A', 'T-C', 'T-G', 'T-T', 'T-GAP', 'GAP-A', 'GAP-C', 'GAP-G', 'GAP-T']\n",
    "\n",
    "    for i in range(0, 24):\n",
    "        if i < 5:\n",
    "            sns.distplot(result5w['astats'][i], color='k', ax=ax[0][i])\n",
    "            ax[0][i].set_title(titles[i], fontsize=17)\n",
    "            ax[0][i].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        elif i > 4 and i < 10:\n",
    "            sns.distplot(result5w['cstats'][i-5], color='k', ax=ax[1][i-5])\n",
    "            ax[1][i-5].set_title(titles[i], fontsize=17)\n",
    "            ax[1][i-5].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        elif i > 9 and i < 15:\n",
    "            sns.distplot(result5w['gstats'][i-10], color='k', ax=ax[2][i-10])\n",
    "            ax[2][i-10].set_title(titles[i], fontsize=17)\n",
    "            ax[2][i-10].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        elif i > 14 and i < 20:\n",
    "            sns.distplot(result5w['tstats'][i-15], color='k', ax=ax[3][i-15])\n",
    "            ax[3][i-15].set_title(titles[i], fontsize=17)\n",
    "            ax[3][i-15].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        else:\n",
    "            sns.distplot(result5w['gapstats'][i-20], color='k', ax=ax[4][i-20])\n",
    "            ax[4][i-20].set_title(titles[i], fontsize=17)\n",
    "            ax[4][i-20].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "    fg.savefig(filename[:-3]+'ind_stats.png', bbox_inches='tight')\n",
    "    return filename[:-3]+'ind_stats.png'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_stats_ind('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.filtered_base_2000-4000.fasta.mothur5.ITS.pcr.ITS.dataframe.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_stats_gen(filename):\n",
    "    frame = pd.DataFrame.from_csv(filename)\n",
    "    result5w = {}\n",
    "    result5w_key_entries = ['expec', 'ostats', 'omatches']\n",
    "    for item in result5w_key_entries:\n",
    "        result5w[item] = []\n",
    "    result5w['expec'] = [frame['expeca'], frame['expecc'], frame['expecg'], frame['expect']]\n",
    "    for i in range(0, len(frame['aap'])):\n",
    "        result5w['omatches'].append(frame['aap'][i]+frame['ccp'][i]+frame['ggp'][i]+frame['ttp'][i])\n",
    "    result5w['ostats'] = [result5w['omatches'], frame['aap'], frame['ccp'], frame['ggp'], frame['ttp']]\n",
    "    \n",
    "    fg, ax = plt.subplots(5, 1, figsize=(20,20), squeeze=False)\n",
    "    fg.suptitle(\"Base Percentages\", fontsize=28, y=1.35)\n",
    "    fg.subplots_adjust(top=1.3)\n",
    "    otitles=['Overall Matches', 'A Matches', 'C Matches', 'G Matches', 'T Matches']\n",
    "\n",
    "    for i in range(0, len(result5w['ostats'])):\n",
    "        sns.distplot(result5w['ostats'][i], color='k', ax=ax[i][0])\n",
    "        ax[i][0].set_title(otitles[i], fontsize=17)\n",
    "        ax[i][0].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        if i > 0:\n",
    "            ax[i][0].set(xlim=(14, 30))\n",
    "            ax[i][0].axvline(np.mean(result5w['expec'][i-1]))\n",
    "    fg.savefig(filename[:-3]+'gen_stats.png', bbox_inches='tight')\n",
    "    return filename[:-3]+'gen_stats.png'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_stats_gen('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.filtered_base_2000-4000.fasta.mothur5.ITS.pcr.ITS.dataframe.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_ITS(fastq, reference_fasta, primer_fasta):\n",
    "    get_fasta(fastq)\n",
    "    print('get_fasta DONE')\n",
    "    get_bedoutput_ITS(reference_fasta, primer_fasta)\n",
    "    print('get_bedoutput_ITS DONE')\n",
    "    get_bedfasta(reference_fasta[:-6]+'.ITS.bedfile', reference_fasta)\n",
    "    print('get_bedfasta DONE')\n",
    "    length_filter(fastq[:-6]+'.fasta', 2000, 4000)\n",
    "    print('length_filter DONE')\n",
    "    pcr_seqs((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta', primer_fasta[:-14]+'.oligos')\n",
    "    print('pcr_seqs DONE')\n",
    "    pairwise_aln(reference_fasta[:-6]+'.ITS.bedoutput.fasta', ((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta')[:-6]+'.mothur5.%s.pcr.fasta' % ((primer_fasta[:-14]+'.oligos')[-10:-7]))\n",
    "    plot_stats_ind(((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta')+'.mothur5.%s.pcr.fasta'[:-5] % ((primer_fasta[:-14]+'.oligos')[-10:-7])+(reference_fasta[:-6]+'.ITS.bedoutput.fasta')[-19:-15]+'dataframe.csv')\n",
    "    plot_stats_gen(((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta')+'.mothur5.%s.pcr.fasta'[:-5] % ((primer_fasta[:-14]+'.oligos')[-10:-7])+(reference_fasta[:-6]+'.ITS.bedoutput.fasta')[-19:-15]+'dataframe.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_ITS('/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.fastq', '/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/zymoseptoria_tritici.reference.fasta', '/media/MassStorage/tmp/TE/summer_project/general_analysis/zymoseptoria_tritici/ITS_primers.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting run_ITS.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile run_ITS.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "def get_fasta(fastq):\n",
    "    SeqIO.convert(fastq, \"fastq\", fastq[:-6]+'.fasta', \"fasta\")\n",
    "    return fastq[:-6]+'.fasta'\n",
    "\n",
    "def get_bedoutput_ITS(reference_fasta, primer_fasta):\n",
    "    \"\"\"The purpose of this function is to generate a fasta file of the reference genome regions between primers, \n",
    "    which should correspond with the region sequenced using the MinION. This requires first the generation of a \n",
    "    database using the nucleotide makeblastdb command, followed by running a blast search across the database \n",
    "    using the primer sequence. These command line commands are processed through subprocess. Subsequently, the \n",
    "    resulting outfile from the blast is read in as a dataframe and this dataframe is manipulated in order to \n",
    "    utilise only the minimum evalues. Each forward sequence is then compared to each reverse sequence, given they\n",
    "    are on the same contig and the resultant read would be within 2000-4000 bp long. These results are saved as \n",
    "    a bedfile for use in the get_bedfasta function\"\"\"\n",
    "    cmd = 'makeblastdb -in %s -dbtype nucl -out %s' % (reference_fasta, reference_fasta[:-6]+'_db')\n",
    "    subprocess.getoutput(cmd)\n",
    "    cmd2 = 'blastn -query %s -db %s -evalue=100000 -task \"blastn-short\" -outfmt 6 > %s' % (primer_fasta, reference_fasta[:-6]+'_db', reference_fasta[:-6]+'.outfmt6')\n",
    "    subprocess.getoutput(cmd2)\n",
    "    df = pd.read_csv(reference_fasta[:-6]+'.outfmt6', sep=\"\\t\", header=None, names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"])\n",
    "    forward_df = df.loc[(df['qseqid']=='forward')]\n",
    "    forward_df = forward_df.loc[df['evalue']==min(forward_df['evalue'])]\n",
    "    forward_df = forward_df.reset_index()\n",
    "    reverse_df = df.loc[(df['qseqid']=='reverse')]\n",
    "    reverse_df = reverse_df.loc[df['evalue']==min(reverse_df['evalue'])]\n",
    "    reverse_df = reverse_df.reset_index()\n",
    "    forward_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "    for i in range(0, len(forward_df)):\n",
    "        if forward_df['sstart'][i] <= forward_df['send'][i]:\n",
    "            forward_bed = forward_bed.append({'chrom': forward_df['sseqid'][i],'chromStart': (forward_df['sstart'][i])-1, 'chromEnd': forward_df['send'][i]}, ignore_index=True)\n",
    "        else:\n",
    "            forward_bed = forward_bed.append({'chrom': forward_df['sseqid'][i],'chromStart': (forward_df['send'][i])-1, 'chromEnd': forward_df['sstart'][i]}, ignore_index=True)\n",
    "    forward_bed.sort_values(['chromStart'])\n",
    "    reverse_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "    for i in reverse_df.index:\n",
    "        if reverse_df['sstart'][i] < reverse_df['send'][i]:\n",
    "            reverse_bed = reverse_bed.append({'chrom': reverse_df['sseqid'][i],'chromStart': (reverse_df['sstart'][i])-1, 'chromEnd': reverse_df['send'][i]}, ignore_index=True)\n",
    "        else:\n",
    "            reverse_bed = reverse_bed.append({'chrom': reverse_df['sseqid'][i],'chromStart': (reverse_df['send'][i])-1, 'chromEnd': reverse_df['sstart'][i]}, ignore_index=True)\n",
    "    reverse_bed.sort_values(['chromStart'])\n",
    "    intervals_list = []\n",
    "    for row in forward_bed.itertuples(index=True, name='Pandas'):\n",
    "        counter = 0\n",
    "        for rows in reverse_bed.itertuples(index=True, name='Pandas'):\n",
    "            # Here lies the difference between the ITS and TEF versions of this function - for ITS the read length range is 2000-4000\n",
    "            if getattr(row, 'chrom') == getattr(rows, 'chrom') and 2000 < np.absolute(getattr(row, 'chromEnd') - getattr(rows, 'chromStart')) < 4000 and getattr(row, 'chromEnd') > getattr(rows, 'chromStart'):\n",
    "                counter += 1\n",
    "                intervals_list.append([forward_bed['chrom'][getattr(row, 'Index')], reverse_bed['chromStart'][getattr(rows, 'Index')], forward_bed['chromEnd'][getattr(row, 'Index')]])\n",
    "            elif getattr(row, 'chrom') == getattr(rows, 'chrom') and 2000 < np.absolute(getattr(row, 'chromEnd') - getattr(rows, 'chromStart')) < 4000 and getattr(row, 'chromEnd') < getattr(rows, 'chromStart'):\n",
    "                intervals_list.append([forward_bed['chrom'][getattr(row, 'Index')], forward_bed['chromEnd'][getattr(row, 'Index')], reverse_bed['chromStart'][getattr(rows, 'Index')]])\n",
    "        if counter > 1:\n",
    "            print(\"WARNING\")\n",
    "    intervals_frame = pd.DataFrame(data=intervals_list, columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "    print(intervals_frame)\n",
    "    intervals_frame.to_csv(reference_fasta[:-6]+'.ITS.bedfile', sep='\\t', header=False, index=False)\n",
    "#     lines = []\n",
    "#     for i in range(0, len(intervals_frame)):\n",
    "#         lines.append(['%i\\t%i\\t%i' % (intervals_frame['chrom'][i], intervals_frame['chromStart'][i], intervals_frame['chromEnd'][i])]) \n",
    "#     with open(reference_fasta[:-6]+'.ITS.bedfile', 'w') as f:\n",
    "#         for i in range(0, len(lines)):\n",
    "#             f.writelines(lines[i])\n",
    "#             f.writelines('\\n')\n",
    "    return reference_fasta[:-6]+'.ITS.bedfile'\n",
    "\n",
    "def get_bedfasta(bedfile, reference_fasta):\n",
    "    cmd2 = 'bedtools getfasta -fo %s -fi %s -bed %s' % (reference_fasta[:-6]+'.ITS.bedoutput.fasta', reference_fasta, bedfile)\n",
    "    subprocess.getoutput(cmd2)\n",
    "    return reference_fasta[:-6]+'.ITS.bedoutput.fasta'\n",
    "\n",
    "def length_filter(fasta, minimum, maximum):\n",
    "    fasta_dict = SeqIO.to_dict(SeqIO.parse(fasta, \"fasta\"))\n",
    "    short_dict = fasta_dict.copy()\n",
    "    for key in fasta_dict:\n",
    "        if len(fasta_dict[key].seq) not in range(minimum, maximum + 1):\n",
    "            del short_dict[key]\n",
    "    SeqIO.write(short_dict.values(), fasta[:-6]+'.filtered_base_%s-%s.fasta' % (minimum, maximum), \"fasta\")\n",
    "    return fasta[:-6]+'.filtered_base_%s-%s.fasta' % (minimum, maximum)\n",
    "\n",
    "def pcr_seqs(filtered_fasta, oligos):\n",
    "    tmp = SeqIO.to_dict(SeqIO.parse(filtered_fasta, \"fasta\"))\n",
    "    new_name = filtered_fasta+'.mothur5.%s.fasta' % (oligos[-10:-7])\n",
    "    SeqIO.write(tmp.values(), new_name, \"fasta\")\n",
    "    m = Mothur()\n",
    "    m.pcr.seqs(fasta=new_name, oligos=oligos, pdiffs=5, rdiffs=5)\n",
    "    return new_name\n",
    "    \n",
    "def pairwise_aln(reference_bed_output, mothur_query_fasta):\n",
    "    gen_dict = SeqIO.to_dict(SeqIO.parse(reference_bed_output, \"fasta\"))\n",
    "    print(gen_dict)\n",
    "    gen_dict_seqs = []\n",
    "    for key in gen_dict:\n",
    "        gen_dict_seqs.append(gen_dict[key].seq)\n",
    "    print(gen_dict_seqs)\n",
    "    q_dict = SeqIO.to_dict(SeqIO.parse(mothur_query_fasta, \"fasta\"))\n",
    "    error5w = {}\n",
    "    error5w_key_entries = ['keys', 'alnslengths', 'aap', 'acp', 'agp', 'atp', 'agapp', 'cap', 'ccp', 'cgp', 'ctp', 'cgapp', 'gap', 'gcp', 'ggp', 'gtp', 'ggapp', 'tap', 'tcp', 'tgp', 'ttp', 'tgapp', 'gapap', 'gapcp', 'gapgp', 'gaptp', 'expeca', 'expecc', 'expecg', 'expect']\n",
    "    for item in error5w_key_entries:\n",
    "        error5w[item] = []\n",
    "    for key in q_dict:\n",
    "        error5w['keys'].append(key)\n",
    "        alignments = pairwise2.align.globalms(gen_dict_seqs[0], q_dict[key].seq, 1, -1, -1, 0)\n",
    "        total = 0\n",
    "        aa = 0\n",
    "        ac = 0\n",
    "        ag = 0\n",
    "        at = 0\n",
    "        agap = 0\n",
    "        ca =0\n",
    "        cc=0\n",
    "        cg=0\n",
    "        ct=0\n",
    "        cgap=0\n",
    "        ga=0\n",
    "        gc=0\n",
    "        gg=0\n",
    "        gt=0\n",
    "        ggap=0\n",
    "        ta=0\n",
    "        tc=0\n",
    "        tg=0\n",
    "        tt=0\n",
    "        tgap=0\n",
    "        gapa=0\n",
    "        gapc=0\n",
    "        gapg=0\n",
    "        gapt=0\n",
    "        tmpa=0\n",
    "        tmpc=0\n",
    "        tmpg=0\n",
    "        tmpt=0\n",
    "        comparison = alignments[0][0:2]\n",
    "        for i in range(0, len(comparison[0])):\n",
    "            if comparison[0][i] == \"A\":\n",
    "                tmpa += 1\n",
    "                if comparison[1][i] == \"C\":\n",
    "                    ac+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    ag+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    at+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"A\":\n",
    "                    aa+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    agap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"C\":\n",
    "                tmpc+=1\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    ca+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    cg+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    ct+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    cc+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    cgap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"G\":\n",
    "                tmpg+=1\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    ga+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    gc+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    gt+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    gg+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    ggap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"T\":\n",
    "                tmpt+=1\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    ta+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    tc+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    tg+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    tt+=1\n",
    "                    total += 1\n",
    "                else:\n",
    "                    tgap+=1\n",
    "                    total += 1\n",
    "            if comparison[0][i] == \"-\":\n",
    "                if comparison[1][i] == \"A\":\n",
    "                    gapa+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"C\":\n",
    "                    gapc+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"G\":\n",
    "                    gapg+=1\n",
    "                    total += 1\n",
    "                elif comparison[1][i] == \"T\":\n",
    "                    gapt+=1\n",
    "                    total += 1\n",
    "        error5w['acp'].append(100*ac/total)\n",
    "        error5w['agp'].append(100*ag/total)\n",
    "        error5w['atp'].append(100*at/total)\n",
    "        error5w['agapp'].append(100*agap/total)\n",
    "        error5w['aap'].append(100*aa/total)\n",
    "        error5w['cap'].append(100*ca/total)\n",
    "        error5w['cgp'].append(100*cg/total)\n",
    "        error5w['ctp'].append(100*ct/total)\n",
    "        error5w['cgapp'].append(100*cgap/total)\n",
    "        error5w['ccp'].append(100*cc/total)\n",
    "        error5w['gap'].append(100*ga/total)\n",
    "        error5w['gcp'].append(100*gc/total)\n",
    "        error5w['gtp'].append(100*gt/total)\n",
    "        error5w['ggapp'].append(100*ggap/total)\n",
    "        error5w['ggp'].append(100*gg/total)\n",
    "        error5w['tap'].append(100*ta/total)\n",
    "        error5w['tcp'].append(100*tc/total)\n",
    "        error5w['tgp'].append(100*tg/total)\n",
    "        error5w['tgapp'].append(100*tgap/total)\n",
    "        error5w['ttp'].append(100*tt/total)\n",
    "        error5w['gapap'].append(100*gapa/total)\n",
    "        error5w['gapcp'].append(100*gapc/total)\n",
    "        error5w['gapgp'].append(100*gapg/total)\n",
    "        error5w['gaptp'].append(100*gapt/total)\n",
    "        error5w['alnslengths'].append(total)\n",
    "        error5w['expeca'].append(100*tmpa/total)\n",
    "        error5w['expecc'].append(100*tmpc/total)\n",
    "        error5w['expecg'].append(100*tmpg/total)\n",
    "        error5w['expect'].append(100*tmpt/total)\n",
    "    error5w_frame = pd.DataFrame.from_dict(error5w, orient='index', columns=error5w['keys'])\n",
    "    error5w_frame = error5w_frame.transpose()\n",
    "    error5w_frame.to_csv(mothur_query_fasta[:-5]+reference_bed_output[-19:-15]+'dataframe.csv')\n",
    "    return mothur_query_fasta[:-5]+reference_bed_output[-19:-15]+'dataframe.csv'\n",
    "\n",
    "def plot_stats_ind(filename):\n",
    "    \n",
    "    frame = pd.DataFrame.from_csv(filename)\n",
    "    result5w = {}\n",
    "    result5w_key_entries = ['astats', 'cstats', 'gstats', 'tstats', 'gapstats', 'expec', 'ostats']\n",
    "    for item in result5w_key_entries:\n",
    "        result5w[item] = []\n",
    "    result5w['astats'] = [frame['aap'], frame['acp'], frame['agp'], frame['atp'], frame['agapp']]\n",
    "    result5w['cstats'] = [frame['cap'], frame['ccp'], frame['cgp'], frame['ctp'], frame['cgapp']]\n",
    "    result5w['gstats'] = [frame['gap'], frame['gcp'], frame['ggp'], frame['gtp'], frame['ggapp']]\n",
    "    result5w['tstats'] = [frame['tap'], frame['tcp'], frame['tgp'], frame['ttp'], frame['tgapp']]\n",
    "    result5w['gapstats'] = [frame['gapap'], frame['gapcp'], frame['gapgp'], frame['gaptp']]\n",
    "    \n",
    "    fg, ax = plt.subplots(5, 5, figsize=(20,20))\n",
    "    fg.suptitle(\"Base Percentages\", fontsize=28, y=1.35)\n",
    "    fg.subplots_adjust(top=1.3)\n",
    "    titles = ['A-A', 'A-C', 'A-G', 'A-T', 'A-GAP', 'C-A', 'C-C', 'C-G', 'C-T', 'C-GAP', 'G-A', 'G-C', 'G-G', 'G-T', 'G-GAP', 'T-A', 'T-C', 'T-G', 'T-T', 'T-GAP', 'GAP-A', 'GAP-C', 'GAP-G', 'GAP-T']\n",
    "\n",
    "    for i in range(0, 24):\n",
    "        if i < 5:\n",
    "            sns.distplot(result5w['astats'][i], color='k', ax=ax[0][i])\n",
    "            ax[0][i].set_title(titles[i], fontsize=17)\n",
    "            ax[0][i].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        elif i > 4 and i < 10:\n",
    "            sns.distplot(result5w['cstats'][i-5], color='k', ax=ax[1][i-5])\n",
    "            ax[1][i-5].set_title(titles[i], fontsize=17)\n",
    "            ax[1][i-5].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        elif i > 9 and i < 15:\n",
    "            sns.distplot(result5w['gstats'][i-10], color='k', ax=ax[2][i-10])\n",
    "            ax[2][i-10].set_title(titles[i], fontsize=17)\n",
    "            ax[2][i-10].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        elif i > 14 and i < 20:\n",
    "            sns.distplot(result5w['tstats'][i-15], color='k', ax=ax[3][i-15])\n",
    "            ax[3][i-15].set_title(titles[i], fontsize=17)\n",
    "            ax[3][i-15].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        else:\n",
    "            sns.distplot(result5w['gapstats'][i-20], color='k', ax=ax[4][i-20])\n",
    "            ax[4][i-20].set_title(titles[i], fontsize=17)\n",
    "            ax[4][i-20].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "    fg.savefig(filename[:-3]+'ind_stats.png', bbox_inches='tight')\n",
    "    return filename[:-3]+'ind_stats.png'\n",
    "\n",
    "def plot_stats_gen(filename):\n",
    "    frame = pd.DataFrame.from_csv(filename)\n",
    "    result5w = {}\n",
    "    result5w_key_entries = ['expec', 'ostats', 'omatches']\n",
    "    for item in result5w_key_entries:\n",
    "        result5w[item] = []\n",
    "    result5w['expec'] = [frame['expeca'], frame['expecc'], frame['expecg'], frame['expect']]\n",
    "    for i in range(0, len(frame['aap'])):\n",
    "        result5w['omatches'].append(frame['aap'][i]+frame['ccp'][i]+frame['ggp'][i]+frame['ttp'][i])\n",
    "    result5w['ostats'] = [result5w['omatches'], frame['aap'], frame['ccp'], frame['ggp'], frame['ttp']]\n",
    "    \n",
    "    fg, ax = plt.subplots(5, 1, figsize=(20,20), squeeze=False)\n",
    "    fg.suptitle(\"Base Percentages\", fontsize=28, y=1.35)\n",
    "    fg.subplots_adjust(top=1.3)\n",
    "    otitles=['Overall Matches', 'A Matches', 'C Matches', 'G Matches', 'T Matches']\n",
    "\n",
    "    for i in range(0, len(result5w['ostats'])):\n",
    "        sns.distplot(result5w['ostats'][i], color='k', ax=ax[i][0])\n",
    "        ax[i][0].set_title(otitles[i], fontsize=17)\n",
    "        ax[i][0].set(xlabel='percentage of changes', ylabel='frequency')\n",
    "        if i > 0:\n",
    "            ax[i][0].set(xlim=(14, 30))\n",
    "            ax[i][0].axvline(np.mean(result5w['expec'][i-1]))\n",
    "    fg.savefig(filename[:-3]+'gen_stats.png', bbox_inches='tight')\n",
    "    \n",
    "    ostats = [[np.mean(result5w['ostats'][0]), np.median(result5w['ostats'][0])], [np.mean(result5w['ostats'][1]), np.median(result5w['ostats'][1])], [np.mean(result5w['ostats'][2]), np.median(result5w['ostats'][2])], [np.mean(result5w['ostats'][3]), np.median(result5w['ostats'][3])], [np.mean(result5w['ostats'][4]), np.median(result5w['ostats'][4])]]\n",
    "    ostatsframe = pd.DataFrame(data=ostats, columns=['Overall Matches', 'A Matches', 'C Matches', 'G Matches', 'T Matches'], index=['Mean', 'Median'])\n",
    "    ostatsframe.to_csv(filename[:-3]+'base_stats.csv')\n",
    "    return filename[:-3]+'gen_stats.png'\n",
    "\n",
    "def run_ITS(fastq, reference_fasta, primer_fasta):\n",
    "    get_fasta(fastq)\n",
    "    print('get_fasta DONE')\n",
    "    get_bedoutput_ITS(reference_fasta, primer_fasta)\n",
    "    print('get_bedoutput_ITS DONE')\n",
    "    get_bedfasta(reference_fasta[:-6]+'.ITS.bedfile', reference_fasta)\n",
    "    print('get_bedfasta DONE')\n",
    "    length_filter(fastq[:-6]+'.fasta', 2000, 4000)\n",
    "    print('length_filter DONE')\n",
    "    pcr_seqs((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta', primer_fasta[:-14]+'.oligos')\n",
    "    print('pcr_seqs DONE')\n",
    "    pairwise_aln(reference_fasta[:-6]+'.ITS.bedoutput.fasta', ((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta')+'.mothur5.%s.pcr.fasta' % ((primer_fasta[:-14]+'.oligos')[-10:-7]))\n",
    "    print('pairwise alignments DONE')\n",
    "    plot_stats_ind(((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta')+'.mothur5.%s.pcr.fasta'[:-5] % ((primer_fasta[:-14]+'.oligos')[-10:-7])+(reference_fasta[:-6]+'.ITS.bedoutput.fasta')[-19:-15]+'dataframe.csv')\n",
    "    plot_stats_gen(((fastq[:-6]+'.fasta')[:-6]+'.filtered_base_2000-4000.fasta')+'.mothur5.%s.pcr.fasta'[:-5] % ((primer_fasta[:-14]+'.oligos')[-10:-7])+(reference_fasta[:-6]+'.ITS.bedoutput.fasta')[-19:-15]+'dataframe.csv')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
